### PR TITLE
Local LLM agent experimentation

### PR DESCRIPTION
**Pull request by bdzimmer**
Updates for a version of `bank_agent` that uses a local LLM. I've gotten the best results by far with this model: phind-codellama-34b-v2. You can download the quantized version from HuggingFace Hub here: https://huggingface.co/TheBloke/Phind-CodeLlama-34B-v2-GGUF (phind-codellama-34b-v2.Q4_K_M.gguf) and put it in the `MODELS_DIR_PATH` defined in `project`

* [Model File URL](https://huggingface.co/TheBloke/Phind-CodeLlama-34B-v2-GGUF/resolve/main/phind-codellama-34b-v2.Q4_K_M.gguf) 

Additional requirements:

* `CMAKE_ARGS="-DLLAMA_METAL=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.82`
    * The CMAKE_ARGS / FORCE_CMAKE variables are required to build with metal support for running on Apple Silicon GPUs and won't be necessary / may not work on other platforms.

Entry points:

* `python -m examples.bank_agent_llama` - local llama version of `examples.bank_agent`.
    * This follows Justin's bankagent / react_chat closely for the sake of comparison. I've tried to leave comments here and in `neosophia.agents.simplelocal` to highlight the differences, mostly prompt changes and additional formatting reminders.
*  `python -m examples.bank_agent_eval` - run evaluation and compare agents using different LLMs, including OpenAI and local. 
    * An update of note here is that the `system` abstraction now uses the run index as a seed, so you can get a range of behavior out of deterministic local LLMs. The seed is ignored for OpenAI LLMs.
    * I've added additional questions, some of which don't always work with the local LLM.

If you're not running on Apple Silicon, you should probably set `GPU_ENABLE = False`.

### CODE REVIEW

File: src/neosophia/llmtools/dispatch.py [hunk 1]
```diff
@@ -54,45 +63,64 @@ def _param_str(pname: str, pdesc: ParamDesc) -> str:
...
-        if line.startswith(func_prefix):
-            name = line.removeprefix(func_prefix)
-        elif line.startswith(param_prefix):
-            line = line.removeprefix(param_prefix).strip()
-            # word at start of line is parameter name
-            words = line.split()
-            pname = words[0]
-            value = line.removeprefix(pname).strip()
+        print('>', f'`{line}`')
+
```
Comment (bdzimmer) [inline src/neosophia/llmtools/dispatch.py:99]: These changes allow slightly incorrect formatting to still be parsed.

File: examples/bank_agent_llama.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            }
+
+            # ~~~~
+
+            model_path = os.path.join(
+                project.MODELS_DIR_PATH,
+                # 'wizardcoder-python-34b-v1.0.Q4_K_M.gguf'
+                # 'vicuna-13b-v1.5.Q4_K_M.gguf'
+                # 'wizardlm-13b-v1.2.Q4_K_M.gguf'
+                'phind-codellama-34b-v2.Q4_K_M.gguf'
```
Comment (bdzimmer) [inline examples/bank_agent_llama.py:205]: This is the model file to download

**Comment by jmpounders** [comment]
Is `llama_cpp` the only additional package that needs to be installed?  Recommend updating the requirements.txt, esp if a certain version is needed.

**Comment by jmpounders** [comment]
> Is `llama_cpp` the only additional package that needs to be installed? Recommend updating the requirements.txt, esp if a certain version is needed.

Reading the description closer now... can you update the requirements for llama_cpp and still get the right cmake flags, or should that always be a one-off installation?

**Comment by bdzimmer** [comment]
@jmpounders I'll look into it, might have to make an install script since it's setting environment variables but there may be a way to do that in requirements.txt. That being said, might be better to keep the metal flags separate since that's platform specific. And I'm realizing I should have made a note about that somewhere in the PR, heh.

File: examples/bank_agent_eval.py [hunk 3]
```diff
@@ -39,25 +90,83 @@ def main():
...
+                model_path=model_path,
+                n_gpu_layers=10000 if GPU_ENABLE else 0,
+                n_ctx=simplelocal.LLAMA2_MAX_TOKENS,
+                seed=seed  # AFAIK this is the only way to set the seed.
+            )
+
+            llama_model_wrapped = simplelocal.build_llama2_wrapper(
+                llama_model, promptformat, stop)
+
+            question_uid = hash(question) % 1000
```
Comment (jmpounders) [inline examples/bank_agent_eval.py:144]: May not matter in this instance, but `hash` is not generally reproducible so the uid will only unique in a given run.  `hashlib` has alternatives.

Not sure what question_uid is being used for, so feel free to ignore if this comment is not relevant.  I imagine if this were an issue you would have noticed already :-)

**Review by jmpounders — APPROVED** [review]
This looks great!  I went through all of the code.  Was not able to run it locally on my 2018 Intel Mac, even with the cmake metal flag off.  Didn't get any errors, and Python was running the CPUs hard, so I'm not sure what was going on.  It just appeared to hang.  Waited 15-20 mins before killing it.

To that end, we should add something to the documentation or release notes indicating that this can only be run on certain hardware at this point.  If we need to, we can further debug my situation to see if we can make it work on older, non-GPU hardware, but that seems like a lower priority in my opinion.

Nice work!

File: examples/bank_agent_eval.py (hunk 3 above)
Comment (bdzimmer) [reply examples/bank_agent_eval.py:144]: Ah nice catch. I thought "hash" did something sensible for strings but I guess not. Was just using this so the saved prepared prompts were unique. I can replace with something better.

**Review by kevindewalt — APPROVED** [review]

**Pull request merged**